### PR TITLE
Complete local quote

### PR DESCRIPTION
**Pull request by ngeiswei**
For issue #980 

It introduces a `Quotation` class to keep track of the quotation state and simplify a lot of the existing code with it while supporting `LocalQuoteLink` as well.

Note that it might slow down a bit the atomspace and the pattern matcher, although I didn't measure a substantial slowdown when running the utest before and after.

@linas I'm confident, but I think it's wiser if you review it.

### CODE REVIEW

File: opencog/atoms/pattern/PatternLink.cc [hunk 1]
```diff
@@ -867,43 +862,18 @@ void PatternLink::make_term_tree_recursive(const Handle& root,
...
-	//
-	// Ignore quoting and unquoting nodes in the PatternTerm
-	if ((not parent->isQuoted() and (QUOTE_LINK == t or LOCAL_QUOTE_LINK == t))
-	    or (parent->getQuotationLevel() == 1
-	        and not parent->isLocalQuoted() and UNQUOTE_LINK == t)) {
-		if (1 != h->getArity())
-			throw InvalidParamException(TRACE_INFO,
-			                            "QuoteLink/UnquoteLink has "
-			                            "unexpected arity!");
-		h = h->getOutgoingAtom(0);
```
Comment (linas) [inline opencog/atoms/pattern/PatternLink.cc]: It seems this line of code was cut, and I don't see what replaces it.  I assume the unit tests still pass, so it is surprising that cutting this still allows things to work ...

Comment (linas) [inline opencog/atoms/pattern/PatternLink.cc]: Never mind. I spoke too soon. I see that it was moved to the PTM. That is.. unexpected, but I guess its OK.

File: opencog/query/DefaultPatternMatchCB.cc [hunk 2]
```diff
@@ -389,33 +391,34 @@ bool DefaultPatternMatchCB::is_self_ground(const Handle& ptrn,
...
 		// valid way to get to here.
-		if (ptype == QUOTE_LINK and 0 == quote_level and ptrn == grnd) return true;
-		if (ptype == UNQUOTE_LINK and 1 == quote_level and ptrn == grnd) return true;
-
-		if (ptype == QUOTE_LINK) quote_level++;
-		else quote_level--;
+		if (quotation.consumable(ptype) and ptrn == grnd) return true;
 
 		const Handle& qpat = ptrn->getOutgoingAtom(0);
-		return is_self_ground(qpat, grnd, term_gnds, varset, quote_level);
```
Comment (linas) [inline opencog/query/DefaultPatternMatchCB.cc]: This is confusing. Up above, the unwrapping of the quote link was done in the pattern term (line 49) -- while here, it is unwrapped a second time?  Or maybe the outgoing set is no longer the outgoing set of the atom, but is instead the outgoing set of the quote?  

That is, it seems that `ptrn->getOutgoingAtom(0)`  has the same value as `ptrn->getHandle()` whenever the ptrn is a QuoteLink. This seems like a kind-of schizo behavior in the code ...

Comment (ngeiswei) [inline opencog/query/DefaultPatternMatchCB.cc]: I would agree, except that it's not a pattern term that is being used here, so I presume the unwrapping didn't apply. As you said pattern terms should be used instead, but that's another, though related, issue.

Comment (linas) [inline opencog/query/DefaultPatternMatchCB.cc]: OK, yes, I see now.  It seems weird to me that the quote-unwrapping happens in the PTM ctor, but perhaps this is dictated by local-quote, and perhaps we've spent enough time on this already.

Comment (ngeiswei) [inline opencog/query/DefaultPatternMatchCB.cc]: It's not specifically due to the local quote, I just felt that this is where it belongs. But this is perhaps more a matter of taste than anything.

**Pull request merged**